server/odc-service/src/main/java/com/oceanbase/odc/service/state/StateRouteAspect.java

```diff
@@ -0,0 +1,141 @@
+        Method method = signature.getMethod();
+        StatefulRoute statefulRoute = findAnnotation(method, StatefulRoute.class);
+        boolean nodeChanged = false;
+        StateManager stateManager = null;
+        RouteInfo routeInfo = null;
+        if (statefulRoute != null) {
+            Object stateIdBySePL = parseStateIdFromParameter(proceedingJoinPoint, statefulRoute.stateIdExpression());
+            stateManager = getStateManager(statefulRoute);
+            routeInfo = stateManager.getRouteInfo(stateIdBySePL);
+            if (!routeInfo.isCurrentNode(properties)) {
+                if (routeHealthManager.isHealthy(routeInfo)) {
+                    DispatchResponse dispatchResponse =
+                            requestDispatcher.forward(routeInfo.getHostName(), routeInfo.getPort());
+                    logTrace(method, stateIdBySePL, routeInfo);
+                    StateRouteFilter.getContext().setDispatchResponse(dispatchResponse);
+                    return null;
+                } else {
+                    nodeChanged = true;
+                    stateManager.preHandleBeforeNodeChange(proceedingJoinPoint, routeInfo);
+                }
+
+            }
+        }
+        Object proceed = proceedingJoinPoint.proceed();
+        if (nodeChanged) {
+            Preconditions.checkNotNull(stateManager);
+            stateManager.afterHandleBeforeNodeChange(proceedingJoinPoint, routeInfo);
+        }
+        return proceed;
+    }
+
+    private void logTrace(Method method, Object stateId, RouteInfo dispatchTo) {
+        if (log.isDebugEnabled()) {
+            String currentNodeHostName = SystemUtils.getHostName();
+            String currentNodeIpAddress =
+                    properties.getOdcHost() == null ? SystemUtils.getLocalIpAddress() : properties.getOdcHost();
+            log.debug(
+                    "current requestId={},traceId={},spanId={},method={},hostName={},ipAddress={}, dispatch to host={}",
+                    TraceContextHolder.getRequestId(), TraceContextHolder.getTraceId(), method.getName(), stateId,
+                    currentNodeHostName, currentNodeIpAddress, dispatchTo);
+        }
+    }
+
+    @NotNull
+    private StateManager getStateManager(StatefulRoute statefulRoute) {
+        StateManager stateManager = null;
+        if (statefulRoute.stateName() != StateName.NONE) {
+            StateName stateName = statefulRoute.stateName();
+            stateManager = SpringContextUtil.getBean(stateName.getStateManagerClass());
+        }
+        if (StringUtils.isNotBlank(statefulRoute.stateManager())) {
+            Object bean = SpringContextUtil.getBean(statefulRoute.stateManager());
+            Preconditions.checkArgument(bean instanceof StateManager, "illegal stateManager type");
+            stateManager = (StateManager) bean;
+        }
+        Preconditions.checkNotNull(stateManager, "stateManager");
```

Comment:
PreCondition is often used for parameter check , here Verify may more suitable

Reply:
done!